### PR TITLE
Flow-based infrastructure reconciliation without Terraformer

### LINKED ISSUE #362: Terraform removal for infrastructure controller
**How to categorize this issue?**
/kind technical-debt
/priority 2
/platform openstack

**What would you like to be added**:
Project Gardener is using Terraform since beginning to create, manage and delete the underlying infrastructure for Shoot clusters (e.g. VPCs, subnets, nat gateways etc.).

As Terraform was great in the beginning to get started, we today experience more and more issues with it. 
For example:
- Less control over the `Infrastructure` creation and deletion process
- Therefore inflexibility to create infrastructure remedies if they are required (depend on upstream Terraform fixes)
- Inability to import resources in the Terraform state in the Gardener context (Terraformer does not support this)
- Terraform provider version incompatibilities
- State lost issues

Therefore we wanted to start dismantling the Terraform usage in this extension and start to adopt provider specific Go SDKs as replacement.

Ideally all Terraform managed resources should be already tracked in the status section of the `Infrastructure` resource.
In case there are resources not tracked in the status then it need to be populated first before the Go SDK clients can take over.

This issue is meant to track progress of the Terraform removal on this extension.

**Why is this needed**:
See above.

/roadmap cloud

### PR DESCRIPTION
**Pull request by MartinWeindel**
**How to categorize this PR?**

/area control-plane
/kind enhancement
/platform openstack

**What this PR does / why we need it**:
Reconciliation and deletion of infrastructure resources are managed directly with a graph flow and using Openstack SDK client API instead of using Terraformer.
Migration for existing infrastructure resources is implemented, too.

**Which issue(s) this PR fixes**:
Fixes #362

**Special notes for your reviewer**:
Currently, to activate the flow based infra management, the shoot or the infrastructure resource must be annotated with
`openstack.provider.extensions.gardener.cloud/use-flow=true`

**Release note**:

```feature operator
Flow-based infrastructure reconciliation without Terraformer
```

### CODE REVIEW

**Comment by MartinWeindel** [comment]
rebased

**Comment by MartinWeindel** [comment]
rebased and addressed review

File: pkg/controller/infrastructure/infraflow/reconcile.go [hunk 1]
```diff
@@ -0,0 +1,468 @@
...
+		Timeout(defaultTimeout), Dependencies(ensureRouter))
+
+	_ = c.AddTask(g, "ensure ssh key pair",
+		c.ensureSSHKeyPair,
+		Timeout(defaultTimeout), Dependencies(ensureRouter))
+
+	return g
+}
+
+func unused(_ any) {}
```
[the remark this answers is not in the document]
Comment (MartinWeindel) [reply pkg/controller/infrastructure/infraflow/reconcile.go]: removed

File: pkg/controller/infrastructure/infraflow/persistent_state.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+import (
+	"fmt"
+
+	"github.com/gardener/gardener-extension-provider-openstack/pkg/controller/infrastructure/infraflow/shared"
+	"k8s.io/apimachinery/pkg/util/json"
+)
+
+const (
+	// PersistentStateVersion is the current version used for persisting the state.
+	PersistentStateVersion = "1.0"
```
[the remark this answers is not in the document]
Comment (MartinWeindel) [reply pkg/controller/infrastructure/infraflow/persistent_state.go]: done

File: pkg/controller/infrastructure/actuator.go [hunk 3]
```diff
@@ -58,13 +67,32 @@ func (a *actuator) updateProviderStatus(
...
 	return a.Client().Status().Patch(ctx, infra, patch)
 }
+
+func (a *actuator) addErrorCodes(err error) error {
+	if err == nil {
+		return nil
+	}
+	if msg := err.Error(); strings.Contains(msg, "PolicyNotAuthorized") {
+		return helper.NewErrorWithCodes(err, gardencorev1beta1.ErrorInfraUnauthorized)
+	}
```
[the remark this answers is not in the document]
Comment (MartinWeindel) [reply pkg/controller/infrastructure/actuator.go:96]: done

File: pkg/controller/infrastructure/infraflow/access/networking_access.go [hunk 4]
```diff
@@ -0,0 +1,526 @@
...
+	r, err := regexp.Compile(floatingPoolSubnetNameRegex)
+	if err != nil {
+		return nil, err
+	}
+	for _, subnet := range allSubnets {
+		// Check for a very rare case where the response would include no
+		// subnet name. No name means nothing to attempt a match against,
+		// therefore we are skipping such subnet.
+		if subnet.Name == "" {
+			a.log.Info("[WARN] Unable to find subnet name to match against for subnet ID, nothing to do.",
```
[the remark this answers is not in the document]
Comment (MartinWeindel) [reply pkg/controller/infrastructure/infraflow/access/networking_access.go]: using `a.log.V(2).Info`

File: pkg/controller/infrastructure/infraflow/reconcile.go [hunk 5]
```diff
@@ -0,0 +1,484 @@
...
+
+func (c *FlowContext) findExistingRouter() (*access.Router, error) {
+	return findExisting(c.state.Get(IdentifierRouter), c.namespace, c.access.GetRouterByID, c.access.GetRouterByName)
+}
+
+func (c *FlowContext) getRouterID() (*string, error) {
+	if c.config.Networks.Router != nil {
+		return &c.config.Networks.Router.ID, nil
+	}
+	routerID := c.state.Get(IdentifierRouter)
```
[the remark this answers is not in the document]
Comment (MartinWeindel) [reply pkg/controller/infrastructure/infraflow/reconcile.go]: removed methods `getExistingRouterAndSubnetIDs`, `getRouterID`, `getSubnetID` and using recover steps in deletion flow

File: pkg/controller/infrastructure/infraflow/reconcile.go [hunk 6]
```diff
@@ -0,0 +1,484 @@
...
+		}); err != nil {
+			return err
+		} else if modified {
+			log.Info("updated rules")
+		}
+	} else {
+		log.Info("creating...")
+		created, err := c.access.CreateSecurityGroup(desired)
+		if err != nil {
+			return err
```
[the remark this answers is not in the document]
Comment (MartinWeindel) [reply pkg/controller/infrastructure/infraflow/reconcile.go]: done

File: pkg/controller/infrastructure/infraflow/delete.go [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+	_ = c.AddTask(g, "delete router",
+		c.deleteRouter,
+		DoIf(needToDeleteRouter), Timeout(defaultTimeout), Dependencies(deleteRouterInterface))
+
+	return g
+}
+
+func (c *FlowContext) hasRouter() bool {
+	return !c.state.IsAlreadyDeleted(IdentifierRouter)
+}
```
[the remark this answers is not in the document]
Comment (MartinWeindel) [reply pkg/controller/infrastructure/infraflow/delete.go]: removed

File: pkg/controller/infrastructure/infraflow/delete.go [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+		return err
+	}
+	return nil
+}
+
+func (c *FlowContext) deleteSecGroup(ctx context.Context) error {
+	if c.state.IsAlreadyDeleted(IdentifierSecGroup) {
+		return nil
+	}
+
```
[the remark this answers is not in the document]
Comment (MartinWeindel) [reply pkg/controller/infrastructure/infraflow/delete.go]: dropped usage of `IsAlreadyDeleted` and `SetDeleted`

File: pkg/openstack/client/networking.go [hunk 9]
```diff
@@ -169,3 +218,71 @@ func (c *NetworkingClient) UpdateRoutesForRouter(routes []routers.Route, routerI
...
+}
+
+// GetRouterInterfacePort gets a port for a router interface
+func (c *NetworkingClient) GetRouterInterfacePort(routerID string) (*ports.Port, error) {
+	page, err := ports.List(c.client, ports.ListOpts{
+		DeviceOwner: "network:router_interface",
+		DeviceID:    routerID,
+	}).AllPages()
+	if err != nil {
+		return nil, err
```
[the remark this answers is not in the document]
Comment (MartinWeindel) [reply pkg/openstack/client/networking.go]: done

File: pkg/controller/infrastructure/infraflow/access/retry.go [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+	NeutronError neutronError
+}
+
+type neutronError struct {
+	Message string `json:"message"`
+	Type    string `json:"type"`
+	Detail  string `json:"detail"`
+}
+
+func retryOn409(log logr.Logger, err error) bool {
```
[the remark this answers is not in the document]
Comment (MartinWeindel) [reply pkg/controller/infrastructure/infraflow/access/retry.go]: renamed

File: pkg/controller/infrastructure/infraflow/reconcile.go [hunk 11]
```diff
@@ -0,0 +1,457 @@
...
+	g := c.buildReconcileGraph()
+	f := g.Compile()
+	if err := f.Run(ctx, flow.Opts{Log: c.Log}); err != nil {
+		return flow.Causes(err)
+	}
+	return nil
+}
+
+func (c *FlowContext) buildReconcileGraph() *flow.Graph {
+	//createNetwork := c.config.Networks.ID == nil
```
Comment (kon-angelo) [inline pkg/controller/infrastructure/infraflow/reconcile.go:48]: This can be removed

**Review by kon-angelo — COMMENTED** [review]
Otherwise lgtm

File: pkg/controller/infrastructure/infraflow/reconcile.go (hunk 11 above)
Comment (MartinWeindel) [reply pkg/controller/infrastructure/infraflow/reconcile.go]: Oops, I removed it. Thanks!

**Review by kon-angelo — APPROVED** [review]
/lgtm

**Pull request merged**